File: JavascriptAlgorithm/Tree/isValidBST-098.js
var isValidBST = function (root) {
  // left  root  right
  let res = [];
  const dfs = (node) => {
    if (node === null) {
      return true;
    }
    let left = dfs(node.left);
    if (res.length > 0 && node.val <= res[res.length - 1]) {
      return false;
    } else {
      res.push(node.val);
    }
    let right = dfs(node.right);
    return left && right;
  };
  return dfs(root);
};

export { isValidBST };

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SecondRound/Chapter07/isValidBST_098.java
package algo.tzashinorpu.SecondRound.Chapter07;

import algo.tzashinorpu.SecondRound.TreeNode;

public class isValidBST_098 {
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean isValidBST(TreeNode root, long lower, long upper) {
        if (root == null) {
            return true;
        }
        if (root.val <= lower) {
            return false;
        }
        if (root.val >= upper) {
            return false;
        }
        return isValidBST(root.left, lower, root.val) && isValidBST(root.right, root.val, upper);
    }
}
